### ISSUE TITLE
hk.multi_transform and jax.grads

### ISSUE THREAD

**Issue by wbrenton** [deepmind/dm-haiku#530]
I'm curious how using hk.multi_transform affects grads tracked by the jax.grad function.

Using the example provided in the documentation for hk.multi_transform, say that you only wanted to update the encoder params. 

You do something like
''' 
encoder_apply, decoder_apply = f.apply

def encoder_loss(params, inputs, labels):
     pred = encoder_apply(params, rng, inputs)
     return ((preds - labels)**2)

grads = jax.grad(encoder_loss)(params, inputs, labels)
# apply grads and get new params
'''
my question is: what are the grads for the decoder params (inside of the params variable, i.e. name='params/decoder/...')? My hope is that they are all zeros (params + lr * grads) has no affect on params.

At the end of the day my goal is to have one unified network with sub-architectures (e.g. encoder-decoder) and on some training iterations only update the encoder params, and on other only update the decoder params.

Is hk.multitransform the right way to go about partitioning the params while still maintaining the ability to use the network as a whole? If not any suggestion on what might be a better route.

Thank you in advance

**Comment by wbrenton** [comment]
Minimal example below. The grads for the subarchitecture not being used are indeed zero. Makes perfect sense being that the decoder wouldn't be tracked on the computation graph used by jax.grad

`def f():
  encoder = hk.Linear(1, name="encoder")
  decoder = hk.Linear(1, name="decoder")

  def init(x):
    z = encoder(x)
    return decoder(z)

  return init, (encoder, decoder)

f = hk.multi_transform(f)
rng = jax.random.PRNGKey(42)
x = jnp.ones([1, 1])
params = f.init(rng, x)

encoder_apply, decoder_apply = f.apply

def loss(params, inputs, labels):
    preds = encoder_apply(params, rng, inputs)
    loss = (preds - labels)**2
    return loss.mean()

loss, grads = jax.value_and_grad(loss)(params, jnp.full([1,1], 1), jnp.full([1,1], 10))

grads`

>>> {'decoder': {'b': DeviceArray([0.], dtype=float32),
  'w': DeviceArray([[0.]], dtype=float32)},
 'encoder': {'b': DeviceArray([-17.485035], dtype=float32),
  'w': DeviceArray([[-17.485035]], dtype=float32)}}
